beego/adapter/config.NewConfig() crashed with an invalid Yamlcontext

Please answer these questions before submitting your issue. Thanks!

1. What version of Go and beego are you using (`bee version`)?
go version go1.16.3 linux/amd64，
with beego commit id: 12af439a9c487f8ce3c561f4fe669f3ef2b3b0a4(latest at 2021/07/19)

2. What operating system and processor architecture are you using (`go env`)?
<details><summary><code>go env</code> </summary><br><pre>
$ go env
GO111MODULE="on"
GOARCH="amd64"
GOBIN="/home/lyf/workspace/gowork/bin"
GOCACHE="/home/lyf/.cache/go-build"
GOENV="/home/lyf/.config/go/env"
GOEXE=""
GOFLAGS="-mod=mod"
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GOINSECURE=""
GOMODCACHE="/home/lyf/workspace/gowork/pkg/mod"
GONOPROXY=""
GONOSUMDB=""
GOOS="linux"
GOPATH="/home/lyf/workspace/gowork"
GOPRIVATE=""
GOPROXY="https://goproxy.cn,direct"
GOROOT="/home/lyf/.local/go"
GOSUMDB="sum.golang.org"
GOTMPDIR=""
GOTOOLDIR="/home/lyf/.local/go/pkg/tool/linux_amd64"
GOVCS=""
GOVERSION="go1.16.3"
GCCGO="gccgo"
AR="ar"
CC="gcc"
CXX="g++"
CGO_ENABLED="1"
GOMOD="/home/lyf/workspace/gowork/src/hwprojects/beego/go.mod"
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build2259273923=/tmp/go-build -gno-record-gcc-switches"
</pre></details>


3. What did you do?
I tried to create a config with xmlcontext, which is invalid. Here is my code:
<details><summary>beego/adapter/config/yaml/Poc.go</summary><br><pre>
<code>
package yaml

import (
	"os"

	"github.com/beego/beego/v2/adapter/config"
)

func PoC(data []byte) {
	var (
		yamlcontext = string(data)
	)
	cfgFileName := "testyaml.conf"
	f, err := os.Create(cfgFileName)
	if err != nil {
		panic(err)
	}
	_, err = f.WriteString(yamlcontext)
	if err != nil {
		f.Close()
		panic(err)
	}
	f.Close()
	defer os.Remove(cfgFileName)
	yamlconf, err := config.NewConfig("yaml", cfgFileName)
	if err != nil && yamlconf != nil {
		panic(err)
	} else if err != nil {
		return
	}
}
</code>
</details>

<details><summary>beego/adapter/config/yaml/main/main.go</summary><br><pre>
<code>
package main

import (
	"github.com/beego/beego/v2/adapter/config/yaml"
)

func main() {

	data := []byte("{\x00:")
	yaml.PoC(data)
}
</code>
</details>


4. What did you expect to see?
Create the config according to xmlcontext successfully or fail with an err.

5. What did you see instead?
It crashed with error: "panic: runtime error: index out of range [5] with length 5"

It's so strange. It seems that the crash is introduced by an import to "github.com/beego/beego/v2/adapter/config/yaml"

And  this also crashed when "core" package was used

Here is the new poc:

<details><summary>beego/adapter/config/yaml/Poc.go</summary><br><pre>
<code>
package main

import (
	"os"

	"github.com/beego/beego/v2/core/config"
	_ "github.com/beego/beego/v2/core/config/yaml"
)

func main() {

	data := []byte("{\x00:")
	PoC(data)
}

func PoC(data []byte) {
	var (
		yamlcontext = string(data)
	)
	cfgFileName := "testyaml.conf"
	f, err := os.Create(cfgFileName)
	if err != nil {
		panic(err)
	}
	_, err = f.WriteString(yamlcontext)
	if err != nil {
		f.Close()
		panic(err)
	}
	f.Close()
	defer os.Remove(cfgFileName)
	yamlconf, err := config.NewConfig("yaml", cfgFileName)
	if err != nil && yamlconf != nil {
		panic(err)
	} else if err != nil {
		return
	}
}

</code>
</details>

It will crash with error:

2021/07/27 01:19:18.795 [D]  init global config instance failed. If you donot use this, just ignore it.  open conf/app.conf: no such file or directory

2021/07/27 01:19:18 Look like a Json, try json umarshal
panic: runtime error: index out of range [5] with length 5

goroutine 1 [running]:
github.com/beego/goyaml2.(*yamlReader).asMapKeyValue(0xc00015dc00, 0xc000018e43, 0x7, 0xc000018e40, 0x3, 0xc000018e43, 0x7, 0x96eaa0, 0x0)
        /home/jx/workspace/gowork/pkg/mod/github.com/beego/goyaml2@v0.0.0-20130207012346-5545475820dd/reader.go:315 +0x91b
github.com/beego/goyaml2.(*yamlReader).ReadObject(0xc00015dc00, 0x0, 0x1000, 0xc000170000, 0x203000, 0x203000)
        /home/jx/workspace/gowork/pkg/mod/github.com/beego/goyaml2@v0.0.0-20130207012346-5545475820dd/reader.go:64 +0x173
github.com/beego/goyaml2.Read(0x7c93a0, 0xc000101410, 0x200, 0x6f7b20, 0xc0000100b0, 0x7c94a0)
        /home/jx/workspace/gowork/pkg/mod/github.com/beego/goyaml2@v0.0.0-20130207012346-5545475820dd/reader.go:20 +0xaf
github.com/beego/beego/v2/core/config/yaml.parseYML(0xc00016e000, 0x3, 0x200, 0x3, 0x200, 0x0)
        /home/jx/Desktop/crashes/beego/core/config/yaml/yaml.go:103 +0xd6
github.com/beego/beego/v2/core/config/yaml.ReadYmlReader(0x765fe1, 0xd, 0xc00006c540, 0x1, 0x0)
        /home/jx/Desktop/crashes/beego/core/config/yaml/yaml.go:85 +0x8a
github.com/beego/beego/v2/core/config/yaml.(*Config).Parse(0x99cc28, 0x765fe1, 0xd, 0x4, 0xc0001243f8, 0x1, 0x0)
        /home/jx/Desktop/crashes/beego/core/config/yaml/yaml.go:55 +0x39
github.com/beego/beego/v2/core/config.NewConfig(0x76376c, 0x4, 0x765fe1, 0xd, 0x3, 0x0, 0x0, 0xc0000426c0)
        /home/jx/Desktop/crashes/beego/core/config/config.go:226 +0x82
main.PoC(0xc00015df65, 0x3, 0x3)
        /home/jx/Desktop/crashes/yaml/main.go:35 +0x1b4
main.main()
        /home/jx/Desktop/crashes/yaml/main.go:14 +0x4d
exit status 2

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -60,6 +60,7 @@
 - fix bug:reflect.ValueOf(nil) in getFlatParams [4715](https://github.com/beego/beego/pull/4715)
 - Fix 4736: set a fixed value "/" to the "Path" of "_xsrf" cookie. [4736](https://github.com/beego/beego/issues/4735) [4739](https://github.com/beego/beego/issues/4739)
 - Fix 4734: do not reset id in Delete function. [4738](https://github.com/beego/beego/pull/4738) [4742](https://github.com/beego/beego/pull/4742)
+- Fix 4699: Remove Remove goyaml2 dependency. [4755](https://github.com/beego/beego/pull/4755)
 
 ## Fix Sonar
 

--- a/core/config/config.go
+++ b/core/config/config.go
@@ -187,11 +187,11 @@ func (c *BaseConfiger) Strings(key string) ([]string, error) {
 	return strings.Split(res, ";"), nil
 }
 
-func (c *BaseConfiger) Sub(key string) (Configer, error) {
+func (*BaseConfiger) Sub(string) (Configer, error) {
 	return nil, errors.New("unsupported operation")
 }
 
-func (c *BaseConfiger) OnChange(key string, fn func(value string)) {
+func (*BaseConfiger) OnChange(_ string, _ func(value string)) {
 	// do nothing
 }
 
@@ -249,6 +249,12 @@ func ExpandValueEnvForMap(m map[string]interface{}) map[string]interface{} {
 				value[k2] = ExpandValueEnv(v2)
 			}
 			m[k] = value
+		case map[interface{}]interface{}:
+			tmp := make(map[string]interface{}, len(value))
+			for k2, v2 := range value {
+				tmp[k2.(string)] = v2
+			}
+			m[k] = ExpandValueEnvForMap(tmp)
 		}
 	}
 	return m

--- a/core/config/yaml/yaml.go
+++ b/core/config/yaml/yaml.go
@@ -13,11 +13,6 @@
 // limitations under the License.
 
 // Package yaml for config provider
-//
-// depend on github.com/beego/goyaml2
-//
-// go install github.com/beego/goyaml2
-//
 // Usage:
 //  import(
 //   _ "github.com/beego/beego/v2/core/config/yaml"
@@ -30,17 +25,13 @@
 package yaml
 
 import (
-	"bytes"
-	"encoding/json"
 	"errors"
 	"fmt"
 	"io/ioutil"
-	"log"
 	"os"
 	"strings"
 	"sync"
 
-	"github.com/beego/goyaml2"
 	"gopkg.in/yaml.v2"
 
 	"github.com/beego/beego/v2/core/config"
@@ -51,7 +42,7 @@ import (
 type Config struct{}
 
 // Parse returns a ConfigContainer with parsed yaml config map.
-func (yaml *Config) Parse(filename string) (y config.Configer, err error) {
+func (*Config) Parse(filename string) (y config.Configer, err error) {
 	cnf, err := ReadYmlReader(filename)
 	if err != nil {
 		return
@@ -63,7 +54,7 @@ func (yaml *Config) Parse(filename string) (y config.Configer, err error) {
 }
 
 // ParseData parse yaml data
-func (yaml *Config) ParseData(data []byte) (config.Configer, error) {
+func (*Config) ParseData(data []byte) (config.Configer, error) {
 	cnf, err := parseYML(data)
 	if err != nil {
 		return nil, err
@@ -75,7 +66,6 @@ func (yaml *Config) ParseData(data []byte) (config.Configer, error) {
 }
 
 // ReadYmlReader Read yaml file to map.
-// if json like, use json package, unless goyaml2 package.
 func ReadYmlReader(path string) (cnf map[string]interface{}, err error) {
 	buf, err := ioutil.ReadFile(path)
 	if err != nil {
@@ -86,37 +76,14 @@ func ReadYmlReader(path string) (cnf map[string]interface{}, err error) {
 }
 
 // parseYML parse yaml formatted []byte to map.
-func parseYML(buf []byte) (cnf map[string]interface{}, err error) {
-	if len(buf) < 3 {
-		return
-	}
-
-	if string(buf[0:1]) == "{" {
-		log.Println("Look like a Json, try json umarshal")
-		err = json.Unmarshal(buf, &cnf)
-		if err == nil {
-			log.Println("It is Json Map")
-			return
-		}
-	}
-
-	data, err := goyaml2.Read(bytes.NewReader(buf))
+func parseYML(buf []byte) (map[string]interface{}, error) {
+	cnf := make(map[string]interface{})
+	err := yaml.Unmarshal(buf, cnf)
 	if err != nil {
-		log.Println("Goyaml2 ERR>", string(buf), err)
-		return
-	}
-
-	if data == nil {
-		log.Println("Goyaml2 output nil? Pls report bug\n" + string(buf))
-		return
-	}
-	cnf, ok := data.(map[string]interface{})
-	if !ok {
-		log.Println("Not a Map? >> ", string(buf), data)
-		cnf = nil
+		return nil, err
 	}
 	cnf = config.ExpandValueEnvForMap(cnf)
-	return
+	return cnf, err
 }
 
 // ConfigContainer is a config which represents the yaml configuration.
@@ -126,8 +93,8 @@ type ConfigContainer struct {
 }
 
 // Unmarshaler is similar to Sub
-func (c *ConfigContainer) Unmarshaler(prefix string, obj interface{}, opt ...config.DecodeOption) error {
-	sub, err := c.sub(prefix)
+func (c *ConfigContainer) Unmarshaler(prefix string, obj interface{}, _ ...config.DecodeOption) error {
+	sub, err := c.subMap(prefix)
 	if err != nil {
 		return err
 	}
@@ -140,7 +107,7 @@ func (c *ConfigContainer) Unmarshaler(prefix string, obj interface{}, opt ...con
 }
 
 func (c *ConfigContainer) Sub(key string) (config.Configer, error) {
-	sub, err := c.sub(key)
+	sub, err := c.subMap(key)
 	if err != nil {
 		return nil, err
 	}
@@ -149,21 +116,19 @@ func (c *ConfigContainer) Sub(key string) (config.Configer, error) {
 	}, nil
 }
 
-func (c *ConfigContainer) sub(key string) (map[string]interface{}, error) {
+func (c *ConfigContainer) subMap(key string) (map[string]interface{}, error) {
 	tmpData := c.data
 	keys := strings.Split(key, ".")
 	for idx, k := range keys {
 		if v, ok := tmpData[k]; ok {
-			switch v.(type) {
+			switch val := v.(type) {
 			case map[string]interface{}:
-				{
-					tmpData = v.(map[string]interface{})
-					if idx == len(keys)-1 {
-						return tmpData, nil
-					}
+				tmpData = val
+				if idx == len(keys)-1 {
+					return tmpData, nil
 				}
 			default:
-				return nil, errors.New(fmt.Sprintf("the key is invalid: %s", key))
+				return nil, fmt.Errorf("the key is invalid: %s", key)
 			}
 		}
 	}
@@ -171,7 +136,7 @@ func (c *ConfigContainer) sub(key string) (map[string]interface{}, error) {
 	return tmpData, nil
 }
 
-func (c *ConfigContainer) OnChange(key string, fn func(value string)) {
+func (*ConfigContainer) OnChange(_ string, _ func(value string)) {
 	// do nothing
 	logs.Warn("Unsupported operation: OnChange")
 }
@@ -219,12 +184,18 @@ func (c *ConfigContainer) DefaultInt(key string, defaultVal int) int {
 
 // Int64 returns the int64 value for a given key.
 func (c *ConfigContainer) Int64(key string) (int64, error) {
-	if v, err := c.getData(key); err != nil {
+	v, err := c.getData(key)
+	if err != nil {
 		return 0, err
-	} else if vv, ok := v.(int64); ok {
-		return vv, nil
 	}
-	return 0, errors.New("not bool value")
+	switch val := v.(type) {
+	case int:
+		return int64(val), nil
+	case int64:
+		return val, nil
+	default:
+		return 0, errors.New("not int or int64 value")
+	}
 }
 
 // DefaultInt64 returns the int64 value for a given key.
@@ -303,7 +274,18 @@ func (c *ConfigContainer) DefaultStrings(key string, defaultVal []string) []stri
 // GetSection returns map for the given section
 func (c *ConfigContainer) GetSection(section string) (map[string]string, error) {
 	if v, ok := c.data[section]; ok {
-		return v.(map[string]string), nil
+		switch val := v.(type) {
+		case map[string]interface{}:
+			res := make(map[string]string, len(val))
+			for k2, v2 := range val {
+				res[k2] = fmt.Sprintf("%v", v2)
+			}
+			return res, nil
+		case map[string]string:
+			return val, nil
+		default:
+			return nil, fmt.Errorf("unexpected type: %v", v)
+		}
 	}
 	return nil, errors.New("not exist section")
 }
@@ -316,7 +298,11 @@ func (c *ConfigContainer) SaveConfigFile(filename string) (err error) {
 		return err
 	}
 	defer f.Close()
-	err = goyaml2.Write(f, c.data)
+	buf, err := yaml.Marshal(c.data)
+	if err != nil {
+		return err
+	}
+	_, err = f.Write(buf)
 	return err
 }
 
@@ -334,37 +320,28 @@ func (c *ConfigContainer) DIY(key string) (v interface{}, err error) {
 }
 
 func (c *ConfigContainer) getData(key string) (interface{}, error) {
-	if len(key) == 0 {
+	if key == "" {
 		return nil, errors.New("key is empty")
 	}
 	c.RLock()
 	defer c.RUnlock()
 
-	keys := strings.Split(c.key(key), ".")
+	keys := strings.Split(key, ".")
 	tmpData := c.data
 	for idx, k := range keys {
 		if v, ok := tmpData[k]; ok {
-			switch v.(type) {
+			switch val := v.(type) {
 			case map[string]interface{}:
-				{
-					tmpData = v.(map[string]interface{})
-					if idx == len(keys)-1 {
-						return tmpData, nil
-					}
+				tmpData = val
+				if idx == len(keys)-1 {
+					return tmpData, nil
 				}
 			default:
-				{
-					return v, nil
-				}
-
+				return v, nil
 			}
 		}
 	}
 	return nil, fmt.Errorf("not exist key %q", key)
-}
-
-func (c *ConfigContainer) key(key string) string {
-	return key
 }
 
 func init() {

--- a/core/config/yaml/yaml_test.go
+++ b/core/config/yaml/yaml_test.go
@@ -58,6 +58,7 @@ func TestYaml(t *testing.T) {
 			"emptystrings":    []string{},
 		}
 	)
+
 	f, err := os.Create("testyaml.conf")
 	if err != nil {
 		t.Fatal(err)
@@ -69,10 +70,27 @@ func TestYaml(t *testing.T) {
 	}
 	f.Close()
 	defer os.Remove("testyaml.conf")
+
 	yamlconf, err := config.NewConfig("yaml", "testyaml.conf")
 	if err != nil {
 		t.Fatal(err)
 	}
+
+	m, err := ReadYmlReader("testyaml.conf")
+	if err != nil {
+		t.Fatal(err)
+	}
+	assert.Equal(t, m, yamlconf.(*ConfigContainer).data)
+
+	shadow, err := (&Config{}).ParseData([]byte(yamlcontext))
+	if err != nil {
+		t.Fatal(err)
+	}
+	assert.Equal(t, shadow, yamlconf)
+
+	yamlconf.OnChange("abc", func(value string) {
+		fmt.Printf("on change, value is %s \n", value)
+	})
 
 	res, _ := yamlconf.String("appname")
 	if res != "beeapi" {
@@ -103,7 +121,7 @@ func TestYaml(t *testing.T) {
 			value, err = yamlconf.DIY(k)
 		}
 		if err != nil {
-			t.Errorf("get key %q value fatal,%v err %s", k, v, err)
+			t.Errorf("get key %q value fatal, %v err %s", k, v, err)
 		} else if fmt.Sprintf("%v", v) != fmt.Sprintf("%v", value) {
 			t.Errorf("get key %q value, want %v got %v .", k, v, value)
 		}
@@ -119,7 +137,9 @@ func TestYaml(t *testing.T) {
 	}
 
 	sub, err := yamlconf.Sub("user")
-	assert.Nil(t, err)
+	if err != nil {
+		t.Fatal(err)
+	}
 	assert.NotNil(t, sub)
 	name, err := sub.String("name")
 	assert.Nil(t, err)
@@ -142,6 +162,38 @@ func TestYaml(t *testing.T) {
 	assert.Nil(t, err)
 	assert.Equal(t, "tom", user.Name)
 	assert.Equal(t, 13, user.Age)
+
+	// default value
+	assert.Equal(t, "beeapi", yamlconf.DefaultString("appname", "invalid"))
+	assert.Equal(t, "invalid", yamlconf.DefaultString("i-appname", "invalid"))
+	assert.Equal(t, 8080, yamlconf.DefaultInt("httpport", 8090))
+	assert.Equal(t, 8090, yamlconf.DefaultInt("i-httpport", 8090))
+	assert.Equal(t, 3.1415976, yamlconf.DefaultFloat("PI", 3.14))
+	assert.Equal(t, 3.14, yamlconf.DefaultFloat("1-PI", 3.14))
+	assert.True(t, yamlconf.DefaultBool("copyrequestbody", false))
+	assert.True(t, yamlconf.DefaultBool("i-copyrequestbody", true))
+	assert.Equal(t, int64(8080), yamlconf.DefaultInt64("httpport", 8090))
+	assert.Equal(t, int64(8090), yamlconf.DefaultInt64("i-httpport", 8090))
+	assert.Equal(t, "tom", yamlconf.DefaultString("user.name", "invalid"))
+	assert.Equal(t, "invalid", yamlconf.DefaultString("user.1-name", "invalid"))
+	assert.Equal(t, []string{"tom"}, yamlconf.DefaultStrings("strings", []string{"tom"}))
+
+	appName, err := yamlconf.DIY("appname")
+	if err != nil {
+		t.Fatal(err)
+	}
+	assert.Equal(t, "beeapi", appName)
+
+	err = yamlconf.SaveConfigFile(f.Name())
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	section, err := yamlconf.GetSection("user")
+	if err != nil {
+		t.Fatal(err)
+	}
+	assert.Equal(t, "tom", section["name"])
 }
 
 type User struct {

--- a/core/utils/safemap.go
+++ b/core/utils/safemap.go
@@ -18,7 +18,7 @@ import (
 	"sync"
 )
 
-// deprecated
+// Deprecated: using sync.Map
 type BeeMap struct {
 	lock *sync.RWMutex
 	bm   map[interface{}]interface{}

--- a/go.mod
+++ b/go.mod
@@ -3,7 +3,6 @@ module github.com/beego/beego/v2
 go 1.14
 
 require (
-	github.com/beego/goyaml2 v0.0.0-20130207012346-5545475820dd
 	github.com/beego/x2j v0.0.0-20131220205130-a0352aadc542
 	github.com/bradfitz/gomemcache v0.0.0-20190913173617-a41fca850d0b
 	github.com/casbin/casbin v1.9.1
@@ -32,7 +31,6 @@ require (
 	github.com/shiena/ansicolor v0.0.0-20200904210342-c7312218db18
 	github.com/ssdb/gossdb v0.0.0-20180723034631-88f6b59b84ec
 	github.com/stretchr/testify v1.7.0
-	github.com/wendal/errors v0.0.0-20181209125328-7f31f4b264ec // indirect
 	go.etcd.io/etcd/client/v3 v3.5.0
 	golang.org/x/crypto v0.0.0-20210513164829-c07d793c2f9a
 	google.golang.org/grpc v1.38.0

